### PR TITLE
improve `setErrorHandler` example

### PR DESCRIPTION
**Pull request by trim21**
#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [X] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

The example in document will make fastify never send a response if handle throw a error which is not expected error. For example if a handle `throw new Error(...)`. Original document is not clear about this and example doesn't handle it.

Original document make me confused, I assumed that rest of errors will be handled automatically without returning them, but turns out it doesn't.  I'm not sure if "error handler return unhanded error and pop them to parent handler" it expected behavior, but looks like it is, so I create this pr to fix them example. And if it's, I hope the document of `setErrorHandler` could be updated to o, so users can know they can pop error to default error handler. https://www.fastify.io/docs/latest/Reference/Server/#seterrorhandler

like this: `GET /bad-case`

```js
fastify.get('/', function(request, reply) {
  reply.code('bad status code').send({ hello: 'world' })
});

fastify.get('/bad-case', function(request, reply) {
  throw new Error("");
});


fastify.setErrorHandler(function(error, request, reply) {
  if (error instanceof Fastify.errorCodes.FST_ERR_BAD_STATUS_CODE) {
    // Log error
    this.log.error(error);
    // Send error response
    reply.status(500).send({ ok: false });
  }
});
```

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

File: docs/Reference/Errors.md [hunk 1]
```diff
@@ -112,7 +112,11 @@ fastify.setErrorHandler(function (error, request, reply) {
     this.log.error(error)
     // Send error response
     reply.status(500).send({ ok: false })
+    return
   }
+  
+  // pop error to default error handler
+  return error
```
Comment (Fdawgs) [inline docs/Reference/Errors.md:119]:
```suggestion
  throw error
```

File: docs/Reference/Errors.md [hunk 2]
```diff
@@ -112,7 +112,11 @@ fastify.setErrorHandler(function (error, request, reply) {
     this.log.error(error)
     // Send error response
     reply.status(500).send({ ok: false })
+    return
```
Comment (Fdawgs) [inline docs/Reference/Errors.md:115]:
```suggestion
    return reply.status(500).send({ ok: false })
```

**Review by Fdawgs — CHANGES_REQUESTED** [review]

**Comment by trim21** [comment]
@Fdawgs your suggestions doesn't show

**Comment by Fdawgs** [comment]
> @Fdawgs your suggestions doesn't show

Yeah something was up when doing it on mobile.
Will review properly next week when back from leave and at a computer.

**Comment by trim21** [comment]
> > @Fdawgs your suggestions doesn't show
> 
> Yeah something was up when doing it on mobile. Will review properly next week when back from leave and at a computer.

I find your suggestions in notification email

**Comment by trim21** [comment]
@Fdawgs you suggest to replace `return error` with  `throw error`, but I don't think it's a good idea.

Here `return error` will use default error handler to handle error and  send a response, but `throw error` in error handle will cause fastify server exit and throw this error.

**Comment by climba03003** [comment]
`return error` and `throw error` should be the same in custom error handler.

**Comment by trim21** [comment]
> `return error` and `throw error` should be the same in custom error handler.

No, try this:

main.mjs
```js
import Fastify from 'fastify';

const fastify = Fastify();

fastify.get('/', async (request, reply) => {
  throw new Error("tt");
});


fastify.setErrorHandler(function(error, request, reply) {
  throw error;
});

console.log(await fastify.listen({ port: 4000, host: '127.0.0.1' }));
```

**Comment by trim21** [comment]
@climba03003 you need to use async handler

**Comment by trim21** [comment]
```diff
- fastify.get('/bad-case', function(request, reply) {
+ fastify.get('/bad-case', async function(request, reply) {
  throw new Error("bad-case");
});
```

**Comment by trim21** [comment]
this will trigger node:internal/process/promises:288            triggerUncaughtException(err, true /* fromPromise */);

```text
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

Error: err
    at Object.<anonymous> (c:\Users\Trim21\proj\bangumi\graphql\lib\server.ts:66:11)
    at preHandlerCallback (c:\Users\Trim21\proj\bangumi\graphql\node_modules\fastify\lib\handleRequest.js:128:37)
    at preValidationCallback (c:\Users\Trim21\proj\bangumi\graphql\node_modules\fastify\lib\handleRequest.js:112:5)
    at handler (c:\Users\Trim21\proj\bangumi\graphql\node_modules\fastify\lib\handleRequest.js:76:7)
    at handleRequest (c:\Users\Trim21\proj\bangumi\graphql\node_modules\fastify\lib\handleRequest.js:24:5)
    at runPreParsing (c:\Users\Trim21\proj\bangumi\graphql\node_modules\fastify\lib\route.js:528:5)
    at Object.routeHandler (c:\Users\Trim21\proj\bangumi\graphql\node_modules\fastify\lib\route.js:475:7)
    at Router.callHandler (c:\Users\Trim21\proj\bangumi\graphql\node_modules\find-my-way\index.js:413:14)
    at Router.lookup (c:\Users\Trim21\proj\bangumi\graphql\node_modules\find-my-way\index.js:391:17)
    at Server.preRouting (c:\Users\Trim21\proj\bangumi\graphql\node_modules\fastify\fastify.js:773:14)
```

**Comment by trim21** [comment]
> Interesting, but given as the example itself. It does work.

This maybe a bug?

**Comment by climba03003** [comment]
> This may be a bug?

I would say yes, it just fall into some edge case that can actually crash the problem.
It creates an error that never able to be catch by the users.

Let see what the others say. @fastify/core

**Comment by trim21** [comment]
them I think it would be `return error` in this PR, considering throwing error has potential bug. 🤔

File: docs/Reference/Errors.md [hunk 3]
```diff
@@ -111,8 +111,11 @@ fastify.setErrorHandler(function (error, request, reply) {
     // Log error
     this.log.error(error)
     // Send error response
-    reply.status(500).send({ ok: false })
+    return reply.status(500).send({ ok: false })
   }
+  
+  // pop error to default error handler
+  return error
```
Comment (climba03003) [inline docs/Reference/Errors.md:118]: Then, I would probably do in this way to reduce the `PromiseLike` around trip.

```suggestion
    reply.status(500).send({ ok: false })
  } else {
    reply.send(error)
  }
```

**Review by climba03003 — APPROVED** [review]
LGTM.

**Comment by trim21** [comment]
I found this bug, you can throw error from both sync/async handler in async error Handler, this is expected behavior

```js
fastify.setErrorHandler(async function(error, request, reply) {
  if (error instanceof Fastify.errorCodes.FST_ERR_BAD_STATUS_CODE) {
    // Send error response
    return reply.status(500).send({ ok: false });
  }

  throw error;
});
```

But you can't throw error from async handler in sync error handler

BUG:

```js
fastify.get('/bad-case', async function(request, reply) {
  throw new Error("bad-case");
});


fastify.setErrorHandler(function(error, request, reply) {
  if (error instanceof Fastify.errorCodes.FST_ERR_BAD_STATUS_CODE) {
    // Send error response
    return reply.status(500).send({ ok: false });
  }

  throw error;
});
```

File: docs/Reference/Errors.md (hunk 3 above)
Comment (trim21) [reply docs/Reference/Errors.md:118]: This improve perf but it behavior differently, this doesn't pop error to parent error handler, for example an instance error handler, and a sub-instance error handler 🤔 as for document I would suggest just return or throw error.

Comment (climba03003) [reply docs/Reference/Errors.md:118]: `return error` is actually doing the same as `reply.send(error)`. But the action is implicitly done by `fastify`.

See Line 69.
https://github.com/fastify/fastify/blob/90fb718ea8239e59b3abdf6906fb82554d8cc0db/lib/error-handler.js#L64-L71

Comment (trim21) [reply docs/Reference/Errors.md:118]: Oh, yes.

Maybe update document here too https://www.fastify.io/docs/latest/Reference/Server/#seterrorhandler ?

**Pull request merged**